convert2rhel/actions/__init__.py

```diff
@@ -230,63 +285,199 @@ def get_actions(actions_path, prefix):
     return actions
 
 
-def resolve_action_order(potential_actions, failed_actions):
+class Stage:
+    def __init__(self, stage_name, task_header=None, next_stage=None):
+        self.stage_name = stage_name
+        self.task_header = task_header if task_header else stage_name
+        self.next_stage = next_stage
+        self._has_run = False
+
+        python_package = importlib.import_module("convert2rhel.actions.%s" % self.stage_name)
+        self.actions = get_actions(python_package.__path__, python_package.__name__ + ".")
+
+    def check_dependencies(self, previous_stage_actions=None):
+        """
+        Make sure dependencies of this Stage and previous stages are satisfied.
+
+        :raises DependencyError: when there is an unresolvable dependency in
+            the set of actions.
+        """
+        # We want to throw an exception if one of the actions fails to resolve
+        # its deps.  We don't care about the return value here.
+        actions_so_far = list(resolve_action_order(self.actions))
+
+        if self.next_stage:
+            self.next_stage.check_dependencies(actions_so_far)
+
+    def run(self, successes=None, failures=None, skips=None):
+        """
+        Run all the actions in Stage and other linked Stages.
+
+        :keyword successes: Actions which have already run and succeeded.
+        :type: Sequence
+        :keyword failures: Actions which have already run and failed.
+        :type: Sequence
+        :return: 2-tuple consisting of two lists.  One with Actions that
+            have succeeded and one of Actions that have failed.  These
+            lists contain the Actions passed in via successes and failures
+            in addition to any that were run in this :class`Stage`.
+
+        :rtype: FinishedActions
+        .. note:: Success is currently defined as an action whose status after
+            running is WARNING or better (WARNING or SUCCESS) and
+            failure as worse than WARNING (OVERRIDABLE, ERROR, FATAL)
+        """
+        logger.task("Prepare: %s" % self.task_header)
+
+        if self._has_run:
+            raise ActionError("Stage %s has already run." % self.stage_name)
+
+        # Default values and make copies so we don't overwrite callers' data
+        if successes is None:
+            successes = []
+        else:
+            successes = list(successes)
+
+        if failures is None:
+            failures = []
+        else:
+            failures = list(failures)
+
+        if skips is None:
+            skips = []
+        else:
+            skips = list(skips)
```

Comment:
What about making this a bit shorter? I think it looks better and easier to read too

```suggestion
successes = list(successes) if successes else None
failures = list(failures) if failures else None
skips = list(skips) if skips else None
```

Reply:
The actual logic would be:

```python
successes = [] if successes is None else list(successes)
failures = [] if failures is None else list(failures)
skips = [] if skips is None else list(skips)
```

 It is shorter, which I like but is it too much logic for such a short space?  Maybe if I make the comment longer?
```suggestion
# Make a mutable copy of these parameters so we don't overwrite the caller's data.
# If they weren't passed in, default to an empty list.
successes = [] if successes is None else list(successes)
failures = [] if failures is None else list(failures)
skips = [] if skips is None else list(skips)
```
(I would replace the existing comment with this version.)

What do you think?

Reply:
Looks even better to me!